### ISSUE TITLE
help with fixing swat 2

### ISSUE THREAD

**Issue by Volcanotron** [elishacloud/dxwrapper#45]
I was wondering if you could give some advice with improving Swat 2's fps at higher resolutions. 
it seems to be bound to the size of the playable viewport  since i can resize to make it smaller to make the game playable again or make it larger to fit a modern resolution but it causes to be unplayable.

I've used dxwrapper,dgvoodoo and many other things 

heres a video of it in action https://www.youtube.com/watch?v=j0UYJCSgN9M

**Comment by elishacloud** [comment]
I have never played this game but I picked it up from [Steam](https://store.steampowered.com/app/607120/Police_Quest_SWAT_2/).  This game draws most of the screen using GDI, however the large map is drawn using ddraw.  This mix of ddraw and GDI in the same screen is likely why most of the other wrappers don't work.  SWAT 2 only uses a few ddraw features, mainly Blt.  The performance issue seems to be some delay in the Blt calls.  I tried sending the `DDBLT_ASYNC` flag to see if that would speed up the Blt's but it had no effect.

Next I tried using [dd7to9](https://github.com/elishacloud/dxwrapper/wiki/DirectDraw-to-Direct3D9-Conversion) to convert ddraw into Direct3D9.  That seemed to solve the performance issue, as expected.  However, Direct3D9 does not play nice with GDI so the GDI parts of the screen simply turned black.

I tried using several other wrappers, such as [dgVoodoo2](http://dege.freeweb.hu/dgVoodoo2/dgVoodoo2.html), [DDrawCompat](https://github.com/narzoul/DDrawCompat) and [cnc-ddraw](https://github.com/CnCNet/cnc-ddraw/releases).  Most of them just crashed.  I was able to get older versions of DDrawCompat working, sort of.  At least they did not crash.  However, the map was not displayed at all.

So I turned to the next solution.  I used the `dd7to9` code, but output this to GDI.  I had to hard code certain coordinates, and there are still minor glitches, but it seemed to work.  You need to set the compatibility settings on the game to use `Windows XP (Service Pack 2)` and 16-bit `Reduced color mode`.

Download patch: [dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/3775526/dxwrapper.zip)

![image](https://user-images.githubusercontent.com/25625456/67629906-512a5b00-f83b-11e9-9cad-b3dd8dc610c8.png)

**Comment by Volcanotron** [comment]
[dxwrapper-swat.log](https://github.com/elishacloud/dxwrapper/files/3775940/dxwrapper-swat.log)

your new fix doesn't seem to want to work on my end, it crashes when loading the mission. tried on both gog and steam versions.
I really appreciate the help

**Comment by elishacloud** [comment]
It looks like it is failing when trying to create the device context, but I am not sure why.  I added some extra logging to help isolate the issue.  Please try the build below and send me the log file.

Download patch: [dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/3777242/dxwrapper.zip)

**Comment by Volcanotron** [comment]
[dxwrapper-swat.log](https://github.com/elishacloud/dxwrapper/files/3777386/dxwrapper-swat.log)
heres the new log

**Comment by elishacloud** [comment]
OK, I believe I see what the issue is.  It will take longer for a generic fix, but this one should work for now.

Download patch: [dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/3777428/dxwrapper.zip)

**Comment by Volcanotron** [comment]
![Untitled](https://user-images.githubusercontent.com/37210517/67657606-5c9c9400-f9ab-11e9-892e-7929ece6205c.jpg)
seems to be working now 
[dxwrapper-swat.log](https://github.com/elishacloud/dxwrapper/files/3777459/dxwrapper-swat.log)
and heres the new log for it. the gog version seems to load into the level faster than the steam version

**Comment by elishacloud** [comment]
@Volcanotron, here is my first attempt at a generic fix for one of the issues.  Can you test this and let me know if it works for you?

[dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/3800157/dxwrapper.zip)

**Comment by Volcanotron** [comment]
[dxwrapper-swat.log](https://github.com/elishacloud/dxwrapper/files/3800383/dxwrapper-swat.log)
yep seems to working champ

**Comment by elishacloud** [comment]
Thanks @Volcanotron!  Making the other part of this fix generic will take a bit longer, however I can test the rest of the fixes myself.  This fix required help from you since I could not reproduce this issue.

**Comment by elishacloud** [comment]
Closing.  If there is still an issue you can re-open.

**Comment by adamscoble** [comment]
Hey! Sorry to resurrect such an old thread. I came across this trying to get SWAT 2 to work on my PC. The fix you posted on Oct 28, 2019 seems to work great, but I have to drop my resolution to 1920x1080. If I play at 2560x1440, it crashes as soon as it tries to open the game. I can obviously live with the lower resolution if I have to, but I wanted to double check I'm not missing something! (Unfortunately the latest release of dxwrapper doesn't seem to fix the game.) Thanks very much! (It also seems to be running at a reasonably low FPS soon, but I'm not sure if that's just how the game is or not, hehe.)

**Comment by mirh** [comment]
If I had to take a gues...
https://github.com/elishacloud/dxwrapper/wiki/Configuration-(advanced)#ddraw

**Comment by elishacloud** [comment]
@adamscoble, if you are not already, try using the latest released build from [here](https://github.com/elishacloud/dxwrapper/releases).  Just extract the dxwrapper.dll and ddraw.dll files from the release and replace the old files.  Also, make sure you have `Dd7to9` and `DdrawWriteToGDI` options enabled in the ini file and set `DdrawOverrideBitMode = 16`.  If you still have issues after that then send me your log file.

@mirh, this game does not use Direct3D so it should not have any issues with 4k.  But it could be a bug in the game also.

**Comment by adamscoble** [comment]
@elishacloud Thanks for the reply! 

So downloading the latest version and changing just the settings you mentioned crashes when I try to start a campaign (a video plays). Setting the .exe to run in Windows XP (Service Pack 2) compatibility mode results in:
**2560x1440**: 
Crash when attempting to start the actual mission. 
[dxwrapper-swat-2560x1440.log](https://github.com/elishacloud/dxwrapper/files/6675141/dxwrapper-swat-2560x1440.log)

**1920x1080**
_Very_ long load times (might not be able to do anything about this of course)
A blue screen in the actual gameplay area (which is I believe one of the original problems of the thread)
However weirdly it has stopped going through to gameplay now and is crashing the same way that 1440p did.
[dxwrapper-swat-1920x1080.log](https://github.com/elishacloud/dxwrapper/files/6675163/dxwrapper-swat-1920x1080.log)
[dxwrapper-swat-1920x1080-crash.log](https://github.com/elishacloud/dxwrapper/files/6675161/dxwrapper-swat-1920x1080-crash.log)

**Possibly unfixable**
Cutscenes play in widescreen (fill the window), it'd be great if they could play in 4:3 but I'm already incredibly thankful for your work! Haha.

**Comment by adamscoble** [comment]
@elishacloud Sorry for the double message! Just confirming the above is what you meant by logs? Absolutely zero pressure, just wanted to double check!

**Comment by elishacloud** [comment]
Sorry for the delay in responding.  I took at look at your logs ans I believe the issue is the following items:

```log
18840 17:12:35.384 m_IDirectDrawSurfaceX::CopySurface Error: not supported for specified source and destination formats! D3DFMT_X8R8G8B8-->D3DFMT_R5G6B5
13564 17:12:41.631 m_IDirectDrawSurfaceX::CreateDCSurface Error: failed to create compatible DC!
```

Neither of these should be happening with this game.  I will need to add more logging into the module to figure out why that is happening.  I will give you an update when I have time to create the additional logging.

**Comment by adamscoble** [comment]
That's great, thanks very much @elishacloud!

**Comment by sbrohl3** [comment]
Started playing this game after so long, and just wanted to say thanks for the up-to-date fixes! Dropping the resolution to 1920x1080 and installing your dxwrapper fix works!

The only bug from keeping me from playing is on "Riot in the Valley". The game attempts to load resources but then crashes to desktop with a weird message saying something about the M40A1 sniper rifle being used by the United States Marine Corps.
Has anyone else seen this?

**Comment by adamscoble** [comment]
Hey @elishacloud! Just checking in to see if any recent releases address this or if you haven't had time to get to it? The latter is completely understandable, just making sure I didn't miss it!

**Comment by elishacloud** [comment]
This issue is resolved with the latest [release](https://github.com/elishacloud/dxwrapper/releases).  Just enable `Dd7to9`.  You should not need any other setting enabled.

**Comment by elishacloud** [comment]
To get this game working in Windows 11 you can:

1. Use dxwrapper with `Dd7to9` enabled.
2. Rename the "movies" directly or disable movies.  The Movie codex causes a crash on Windows 11.

**Comment by mirh** [comment]
Is [replacing](https://www.swat3lastresort.info/index.php/swat-2-startup-crash-fix) the movies also good enough?
(also.. maybe this is just autosuggestion, but could the issue have anything to do with the forced windowed optimizations?)

**Comment by elishacloud** [comment]
Presumably replacing the movies with new movies with the same file name encoded with a different AVI codex would work, though I did not test it.

The issue is a Windows memory error and has to do with memory management on Windows 11.  Also, SWAT 2 only supports windowed mode, AFAIK.  I don't think any kind of optimizations would help here.

**Comment by mirh** [comment]
The optimizations kick in automatically AFAIK, and W11 specifically enrolled windowed apps. 
But anyhow, if this isn't graphics-related.. I would have guessed that microsoft has a shim for every possible quirk of MM? Or is this specific to VBS/HVCI?

**Comment by elishacloud** [comment]
> I would have guessed that microsoft has a shim for every possible quirk of MM?

Yes, I tried a number of shims and could not find one that solves this.  I do know that at least one of the memory shims (HeapEmulation) is broken on Windows 11 and latest versions of Windows 10.  I searched and could not find anyone else who found a shim that solves this issue.  Anyways, I did not want to spend anymore time on this because I have lots of other things to do.

> Or is this specific to VBS/HVCI?

It could be.  I did try and disable all the memory protection I could from the device Exploit Protection page and none of them helps.  Anyways I had VBS enabled on Windows 10 without an issue.  This issue is specific to Windows 11.
